### ISSUE TITLE
Failing test on armhf architecture

### ISSUE THREAD

**Issue by SnarkBoojum** [fredrik-johansson/arb#150]
Hi,

the Debian build infrastructure detected a problem on armhf, see this [log file](https://tests.reproducible-builds.org/debian/rbuild/unstable/armhf/flint-arb_2.8.1-1.rbuild.log).

I hope that helps,

**Comment by fredrik-johansson** [comment]
I'm afraid it doesn't help much... it needs to be debugged on the same machine to find the cause of the segfault.

**Comment by SnarkBoojum** [comment]
I have an armhf box and I could reproduce the segfault, but what can I do to debug the issue?

**Comment by fredrik-johansson** [comment]
You could run the test program with valgrind and see if that tells you the location of the segfault. Or either use gdb or put in some print statements to narrow it down.

**Comment by SnarkBoojum** [comment]
This is what gdb has to say:

    Starting program: /home/jpuydt/flint-arb-2.8.1/build/arf/test/t-add_si 
    [Thread debugging using libthread_db enabled]
    Using host libthread_db library "/lib/arm-linux-gnueabihf/libthread_db.so.1".
    add_si....
    Program received signal SIGSEGV, Segmentation fault.
    arf_set_ui (v=2147483648, x=0xbefff5c4) at /home/jpuydt/flint-arb-2.8.1/arf.h:466
    466             count_leading_zeros(c, v);

**Comment by fredrik-johansson** [comment]
count_leading_zeros is a macro defined in FLINT, using inline asm on most systems. A guess is that the wrong version is selected on armhf?

**Comment by SnarkBoojum** [comment]
How can I know which version is selected? And what is wrong? The choice, or the result of the choice? I also have problems with flint on armhf... though not exactly the same, see link [build2 on this page](https://tests.reproducible-builds.org/debian/rb-pkg/unstable/armhf/flint.html).

**Comment by fredrik-johansson** [comment]
I can't tell what's wrong... it's strange because it's just a simple invocation with one integer input and one integer output, and count_leading_zeros seems to work elsewhere. Is there something special about the value v=2147483648? That would overflow an slong, but count_leading_zeros takes a ulong as input, so there shouldn't be a problem. Does it work for other values? Does count_leading_zeros segfault if you input 2147483648 in a standalone program?

**Comment by SnarkBoojum** [comment]
The same code compiled under an amd64 and an armhf box:

- needed a -lflint on armhf which wasn't needed on amd64 because of undefined references to __flint_clz_tab ;
- printed 0 on armhf and 32 on amd64.

Here is the code:

    #include <stdio.h>
    
    #include "flint.h"
    
    int
    main (int argc, char *argv[])
    {
      unsigned int c;
    
      count_leading_zeros(c, 2147483648);
    
      printf("%d\n", c);
      
      return 0;
    }

**Comment by SnarkBoojum** [comment]
Perhaps I should report it directly to flint upstream? It really looks highly suspicious.

**Comment by fredrik-johansson** [comment]
Sure, perhaps someone will know how to debug it...

**Comment by latinovic** [comment]
Hi,
arf/test/t-add_si fails with a SIGSEGV on MIPS as well.
Similar behaviour is noticed on arf/test/t-sub_si and arf/test/t-mul_si.

I tried your small example but count_leading_zeros seems to work regularly, as on ARM or AMD64.

If arf/test/t-add_si is built with optimization level O1 or lower, instead of O2, test executes successfully.

I took a look at longlong.h, the same count_leading_zeros implementation is used for MIPS and ARM while i386 and amd64 are using their own asm implementation.
If for i386 we use the same macro that is used for ARM and MIPS, arf/test/t-add_si fails with Segmentation fault as well.

Here is count_leading_zeros macro for non x86* architectures:
```
#define count_leading_zeros(count, x)                        \
  do {                                                                                              \
    mp_limb_t __xr = (x);                                                                       \
    mp_limb_t __a;                                                                                 \
                                                                                                          \
    if (GMP_LIMB_BITS == 32)                                                   \
      {                                                                                             \
        __a = __xr < ((mp_limb_t) 1 << 2*__BITS4)                                      \
          ? (__xr < ((mp_limb_t) 1 << __BITS4) ? 1 : __BITS4 + 1) \
          : (__xr < ((mp_limb_t) 1 << 3*__BITS4) ? 2*__BITS4 + 1         \
          : 3*__BITS4 + 1);                                                                   \
      }                                                                                             \
    else                                                                                               \
      {                                                                                             \
        for (__a = GMP_LIMB_BITS - 8; __a > 0; __a -= 8) \
          if (((__xr >> __a) & 0xff) != 0)                                           \
            break;                                                                                  \
        ++__a;                                                                                      \
      }                                                                                             \
                                                                                                          \
    (count) = GMP_LIMB_BITS + 1 - __a - __flint_clz_tab[__xr >> __a]; \
  } while (0)

```

I would also like to mention that if we use the code from else branch (GMP_LIMB_BITS != 32), count_leading_zeros works correctly and arf/test/t-add_si executes successfully for MIPS32 and MIPS64.


I have investigated this further on MIPS.

```
Program received signal SIGSEGV, Segmentation fault.
arf_set_ui (v=2147483648, x=0x7fff64b4) at /build/flint-arb-dejan/flint-arb-2.8.1/arf.h:466
466	        count_leading_zeros(c, v);
(gdb) bt
#0  arf_set_ui (v=2147483648, x=0x7fff64b4) at /build/flint-arb-dejan/flint-arb-2.8.1/arf.h:466
#1  arf_set_si (v=-2147483648, x=0x7fff64b4) at /build/flint-arb-dejan/flint-arb-2.8.1/arf.h:476
#2  arf_add_si_naive (z=z@entry=0x7fff6534, x=x@entry=0x7fff6514, y=y@entry=-2147483648, prec=prec@entry=593, rnd=rnd@entry=0) at test/t-add_si.c:35
```


The main reason for Segmentation fault is usage of _slt_ instruction (Set on less than (signed)),
while comparing input value 2147483648 with 0x10000 (1 << 2*__BITS4).
As 2147483648 is treated as signed number, the result of comparing is not valid and calculation of index of __flint_clz_tab is wrong (out of range).


  

>  0x00400a38 <+56>:	move	v0,a2
>    0x00400a3c <+60>:	beqz	v0,0x400a94 <arf_add_si_naive+148>
>    0x00400a40 <+64>:	lui	v1,0x1
> ->   0x00400a44 <+68>:	slt	v1,v0,v1
>    0x00400a48 <+72>:	beqz	v1,0x400b1c <arf_add_si_naive+284>
>    0x00400a4c <+76>:	lui	v1,0x100
>    0x00400a50 <+80>:	sltiu	v1,v0,256
>    0x00400a54 <+84>:	bnez	v1,0x400b34 <arf_add_si_naive+308>
>    0x00400a58 <+88>:	li	v1,32
>    0x00400a5c <+92>:	li	v1,24
>    0x00400a60 <+96>:	li	t0,9
>    0x00400a64 <+100>:	srlv	t1,v0,t0
>    0x00400a68 <+104>:	lw	t0,-32704(gp)
>    0x00400a6c <+108>:	addu	t0,t0,t1
>    0x00400a70 <+112>:	li	t1,2
> => 0x00400a74 <+116>:	lbu	t0,0(t0)


If test t-add_si is compiled with lower optimization level, _sltu_ (unsigned) instruction is used and tests executes correctly.


I am not sure why compiler with O2 generate code that includes slt instruction, I have tried to add some casts to ulong in arf_set_si (arf.h), FLINT_ABS macro (flint.h), count_leading_zeros macro (longlong.h), but without success.

Here is arf_set_si function in arf.h:

```
ARF_INLINE void
arf_set_si(arf_t x, slong v)
{
    arf_set_ui(x, FLINT_ABS(v));
    if (v < 0)
        ARF_NEG(x);
}
```

With changes below tests executes successfully.

```
--- arf.h.orig	2016-08-12 14:10:12.885186023 +0000
+++ arf.h	2016-08-12 14:12:21.046118018 +0000
@@ -473,7 +473,10 @@
 ARF_INLINE void
 arf_set_si(arf_t x, slong v)
 {
-    arf_set_ui(x, FLINT_ABS(v));
+    ulong uv = v;
+    if (v < 0)
+	uv = -v;
+    arf_set_ui(x, uv);
     if (v < 0)
         ARF_NEG(x);
 }
```


Also, is possible to use clz builtins ( _builtin_clz() for 32-bits and _builtin_clzl() for 64-bits archs),
it works correctly  on mips32, mips64, i386 and amd64.

**To summarize**: 
- Test  t-add_si fails with Segmentation fault for MIPS, ARM, even for i386 if general count_leading_zeros implementation is used.
- On MIPS, usage of slt (signed) instruction is causing this issue. 
- If t-add_si is compiled with lower optimization level (O0 or O1), this issue does not appear.
- Test t-add_si executes successfully if:
 *  test is compiled with lower optimization level,
 *  else branch (GMP_LIMB_BITS != 32) of count_leading_zeros definition is used,
 *  clz builtins are used in count_leading_zeros implementation.
 * do manual abs instead of FLINT_ABS.

I will try to investigate further, any suggestions are welcome.

**Comment by latinovic** [comment]
I will provide a patch that fixes this issue in flint package.

**Comment by latinovic** [comment]
I had created a fix for this issue using gcc clz builtins in count_leading_zeros implementation, longlong.h in flint package. I tested this fix on mips/mipsel, mips64el,  i386 and amd64.

I am planing to propose this patch to flint, but as I do not have access to arm board I was not able to test it.

@SnarkBoojum, could you plase test this solution on arm?
It is needed to build flint with patch bellow, and after that build flint-arb against this new flint.


```
--- flint-2.5.2.orig/longlong.h
+++ flint-2.5.2/longlong.h
@@ -324,6 +324,39 @@
 
 #endif
 
+/* MIPS */
+#if (defined (__mips__) || defined (__arm__))
+
+#ifdef _LONG_LONG_LIMB
+#define count_leading_zeros(count,x)            \
+  do {                                          \
+    FLINT_ASSERT ((x) != 0);                    \
+    (count) = __builtin_clzll (x);              \
+  } while (0)
+#else
+#define count_leading_zeros(count,x)            \
+  do {                                          \
+    FLINT_ASSERT ((x) != 0);                    \
+    (count) = __builtin_clzl (x);               \
+  } while (0)
+#endif
+
+#ifdef _LONG_LONG_LIMB
+#define count_trailing_zeros(count,x)           \
+  do {                                          \
+    FLINT_ASSERT ((x) != 0);                    \
+    (count) = __builtin_ctzll (x);              \
+  } while (0)
+#else
+#define count_trailing_zeros(count,x)           \
+  do {                                          \
+    FLINT_ASSERT ((x) != 0);                    \
+    (count) = __builtin_ctzl (x);               \
+  } while (0)
+#endif
+
+#endif /* MIPS */
+
 #define udiv_qrnnd_int(q, r, n1, n0, d)                                \
   do {									                                      \
     mp_limb_t __d1, __d0, __q1, __q0, __r1, __r0, __m;			        \
@@ -364,6 +397,7 @@
     (r) = __r0;								                                \
   } while (0)
 
+#ifndef count_leading_zeros
 #define count_leading_zeros(count, x)                        \
   do {									                            \
     mp_limb_t __xr = (x);							                \
@@ -386,9 +420,11 @@
 									                                  \
     (count) = GMP_LIMB_BITS + 1 - __a - __flint_clz_tab[__xr >> __a]; \
   } while (0)
+#endif
 
 #if !(GMP_LIMB_BITS == 64 && defined (__ia64))
 
+#ifndef count_trailing_zeros
 #define count_trailing_zeros(count, x)                 \
   do {									                      \
     mp_limb_t __ctz_x = (x);						          \
@@ -397,6 +433,7 @@
     count_leading_zeros (__ctz_c, __ctz_x & -__ctz_x); \
     (count) = GMP_LIMB_BITS - 1 - __ctz_c;	 \
   } while (0)
+#endif
 
 #endif
 

```

**Comment by SnarkBoojum** [comment]
First I cloned flint-arb, configure, make and make check : failure, as expected.

I patched my system's flint/longlong.h : the patch had fuzz, but that was probably because my system's flint is the latest stable release -- it could still be applied.

Then I retried make clean, make and make check: everything is ok.

Congrats!

**Comment by SnarkBoojum** [comment]
I have bad news : the logs [here](https://buildd.debian.org/status/package.php?p=flint-arb&suite=sid) show that the problem is still there for at least hppa and powerpc (see the links "Build-Attempted").

**Comment by latinovic** [comment]
Hi, the patch does not affect these two arch. I was not able to test it so I could not guarantee  that everything would work correctly.
To apply my fix for these two arch it is needed to extend conditional in the patch (add hppa and powerpc).

**Comment by fredrik-johansson** [comment]
Isn't the root problem that FLINT_ABS relies on undefined behavior? Something like

    #define FLINT_ABS(x) ((slong)(x) < 0 ? ((slong)(-(ulong)(x))) : (slong)(x))

should perhaps be used instead of the current implementation.

**Comment by latinovic** [comment]
@fredrik-johansson I have tests you solution, it fixes the issue on i368 and MIPS.

**Comment by SnarkBoojum** [comment]
Do you have a newer patch?

**Comment by fredrik-johansson** [comment]
I have attempted to fix this in arf.h.
